[TASK] Overhaul data / gettext chapter (#1165)
TYPO3-Documentation/TYPO3CMS-Reference-Typoscript: Documentation/Functions/Data.rst

Data / getText
==============

> ..  confval:: data

Reviewer: Do we really need a confval here? It is more of an introduction.

==========

..  contents::
    :local:

..  _data-type-gettext-applicationcontext:

applicationcontext
------------------

..  versionadded:: 13.0

..  confval:: applicationcontext
    :name: data-applicationcontext

    Returns the current :ref:`application context <t3coreapi:application-context>`
    as string.

..  _data-type-gettext-applicationcontext-example:

Example: Evaluate the current application context
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Evaluate if the current application context is "Production":

..  code-block:: typoscript
    :caption: EXT:site_package/Configuration/TypoScript/setup.typoscript

    if {
        value.data = applicationcontext
        equals = Production
    }

..  _data-type-gettext-cobj:

cObj
----

..  confval:: cObj
    :name: data-cObj

    For :ref:`cobj-content` and :ref:`cobj-records` cObjects that are returned by
    a select query, this returns the row number (1,2,3,...) of the current
    cObject record.

    :typoscript:`parentRecordNumber` is the only key available.

..  _data-type-gettext-cobj-example:

Example: Get the number of the current cObject record
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Get the number of the current cObject record:

Reviewer: This is duplicated to the header and can be removed.

..  code-block:: typoscript
    :caption: EXT:site_package/Configuration/TypoScript/setup.typoscript

    lib.foo.data = cObj : parentRecordNumber

..  _data-type-gettext-context:

context
-------

..  confval:: context
    :name: data-context

>     access values from context API (:ref:`see <t3coreapi:context-api>`).

Reviewer: ```suggestion
    Access values from context API (:ref:`see <t3coreapi:context-api>`).
```

    If a property is an array, it is converted into a comma-separated list.

    Syntax:

    ..  code-block:: none
        :caption: TypoScript Syntax

        lib.foo.data = context:<aspectName>:<propertyName>

..  _data-type-gettext-context-example:

> Example: Retrieve current workspace id

Reviewer: ```suggestion
Example: Retrieve current workspace ID
```

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Retrieve current workspace id:

Reviewer: This is duplicated to the header and can therefore be removed.

..  code-block:: typoscript
    :caption: EXT:site_package/Configuration/TypoScript/setup.typoscript

    page.10 = TEXT
    page.10.data = context:workspace:id
    page.10.wrap = You are in workspace: |

..  _data-type-gettext-current:

current
-------

..  confval:: current
    :name: data-current

    current (gets the "current" value)

    ..  TODO: What is the "current" value? We should explain that.

Example: Get the current value
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Get the current value:

Reviewer: This is duplicated to the header. Can be removed.

..  code-block:: typoscript
    :caption: EXT:site_package/Configuration/TypoScript/setup.typoscript

    lib.foo.data = current

..  _data-type-gettext-date:

date
----

..  confval:: date
    :name: data-date
    :default: :typoscript:`d/m Y`

    Can be used with a colon and :ref:`data-type-date-conf`.

..  _data-type-gettext-date-example:

Example:  Get the current time formatted dd-mm-yy
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Get the current time formatted dd-mm-yy:

Reviewer: This is duplicated.

    lib.foo.data = date : d-m-y

..  _data-type-gettext-db:

DB
--

..  confval:: DB
    :name: data-db
    :Syntax: DB : [table name] : [uid] : [field]

    Value from database. Any record from a table in TCA can be selected here.
    Records marked as deleted will not return any value.

    In contrast to other keys, colons are used here to get deeper into the
    structure, instead of pipes.

..  _data-type-gettext-db-example:

Example: Get a header field value of a record
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Get the value of the header field of record with uid 234 from table
:sql:`tt_content`:

..  code-block:: typoscript
    :caption: EXT:site_package/Configuration/TypoScript/setup.typoscript

    lib.foo.data = DB : tt_content:234:header

Get the value of the header field of a record, whose uid is stored in a GET
parameter `myContentId`:

..  code-block:: typoscript
    :caption: EXT:site_package/Configuration/TypoScript/setup.typoscript

    lib.foo.data.dataWrap = DB : tt_content:{GP : myContentId}:header

..  note::
    It is safe to directly use a client-/user-provided input for the id of a DB

>     record here. The function :php:`ContentObjectRenderer->getData` internally

Reviewer: ```suggestion
    record here. The function :php:`ContentObjectRenderer->getData()` internally
```

>     calls the function :php:`PageRepository->getRawRecord`, which converts the

Reviewer: ```suggestion
    calls the function :php:`PageRepository->getRawRecord()`, which converts the
```

>     parameter to int via :php:`QueryBuilder->createNamedParameter`

Reviewer: ```suggestion
    parameter to int via :php:`QueryBuilder->createNamedParameter()`
```

..  _data-type-gettext-debug-example:

Example: Debug the current root-line
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Outputs the current root-line visually in HTML:

..  code-block:: typoscript
    :caption: EXT:site_package/Configuration/TypoScript/setup.typoscript

    lib.foo.data = debug : rootLine

..  _data-type-gettext-field:

field
-----

..  confval:: field
    :name: data-field
    :Syntax: field : [field name from the current :php:`$cObj->data` array in the cObject, multi-dimensional.]

    This gives read access to the current value of an internal global variable determined by the given key.

    -   As default the :php:`$cObj->data` array is :php:`$GLOBALS['TSFE']->page`
        (record of the current page)

    -   In :ref:`TMENU <tmenu>` :php:`$cObj->data` is set in a loop to the page-record for
        each menu item during its rendering process.

    -   In :ref:`cobj-content` / :ref:`cobj-records` :php:`$cObj->data` is set to
        the actual record

    -   In :ref:`GIFBUILDER <gifbuilder>` :php:`$cObj->data` is set to the data
        :ref:`GIFBUILDER <gifbuilder>` is supplied with.

..  _data-type-gettext-field-example:

> Examples

Reviewer: Do we really need this header here? The anchor above can also point to the first example. This way, we don't have a 5th level of headings.

~~~~~~~~

Example: Get header data
^^^^^^^^^^^^^^^^^^^^^^^^

Get content from :php:`$cObj->data['header']`:

..  code-block:: typoscript
    :caption: EXT:site_package/Configuration/TypoScript/setup.typoscript

    lib.foo.data = field : header

Example: Get data of a field
^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Get content from :php:`$cObj->data['fieldname']['level1']['level2']`:

..  code-block:: typoscript
    :caption: EXT:site_package/Configuration/TypoScript/setup.typoscript

    lib.foo.data = field : fieldname|level1|level2

..  _data-type-gettext-file:

file
----

..  confval:: file
    :name: data-file
    :Syntax: file : [uid] : [property]

>     Retrieves a property from a file object (FAL) by identifying it through its

Reviewer: ```suggestion
    Retrieves a property from a file object (:ref:`FAL <t3coreapi:fal>`) by identifying it through its
```

    :sql:`sys_file` UID. Note that during execution of the :ref:`cobj-files` cObject,
    it is also possible to reference the current file with :confval:`data-current` as UID like
    :typoscript:`file : current : size`.

    The following properties are available: name, uid, originalUid, size, sha1,
    extension, mimetype, contents, publicUrl, modification_date, creation_date

    Furthermore when manipulating references (such as images in content elements
    or media in pages), additional properties are available (not all are
    available all the time, it depends on the setup of *references* of the
    :ref:`cobj-files` cObject): title, description, link, alternative.

>     Additionally, any data in the "sys_file_metadata" table can be accessed too.

Reviewer: ```suggestion
    Additionally, any data in the :sql:`sys_file_metadata` table can be accessed too.
```

    See the :ref:`FILES <cobj-files-examples>` cObject for usage examples.

..  _data-type-gettext-file-example:

Example: Get the size of a file
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Get the file size of the file with the sys\_file UID 17:

..  code-block:: typoscript
    :caption: EXT:site_package/Configuration/TypoScript/setup.typoscript

    lib.foo.data = file : 17 : size

..  _data-type-gettext-flexform:

flexform
--------

..  confval:: flexform
    :name: data-flexform
    :Syntax: flexform : [field containing flexform data].[property of this flexform]

>     Access values from FlexForms, for example inside of :sql:`tt_content` record. In contrast

Reviewer: ```suggestion
    Access values from :ref:`FlexForms <t3coreapi:flexforms>`, for example inside of :sql:`tt_content` record. In contrast
```

(sub-)site.  With :typoscript:`fullRootLine` it is possible to go all the way up to page
tree root. The numbers between square brackets indicate to which page each
value of *pointer* would point to:

..  code-block:: text

    - Page tree root [-2]
      |- 1. page before [-1]
        |- Site root (root template here!) [0]
          |- Here you are! [1]

A "slide" parameter can be added just as for the
:ref:`data-type-gettext-levelfield` property.

..  _data-type-gettext-fullrootline-example:

Example: Get the title of the previous page
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Get the title of the page right before the start of the current website:

..  code-block:: typoscript
    :caption: EXT:site_package/Configuration/TypoScript/setup.typoscript

    lib.foo.data = fullRootLine : -1, title

..  _data-type-gettext-getenv:

getenv
------

..  confval:: getenv
    :name: data-getenv

    Value from PHP environment variables.

    For a cached variation of this feature, please refer to :ref:`getEnv <getenv>`.

..  _data-type-gettext-getenv-example:

Example: Get the HTTP referer
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Get the environment var `HTTP_REFERER`:

Reviewer: ```suggestion
Get the environment variable `HTTP_REFERER`:
```

..  code-block:: typoscript
    :caption: EXT:site_package/Configuration/TypoScript/setup.typoscript

    lib.foo.data = getenv : HTTP_REFERER

..  _data-type-gettext-getindpenv:

getIndpEnv
----------

..  confval:: getIndpEnv

>     :name: data-getgetIndpEnvenv

Reviewer: ```suggestion
    :name: data-getIndpEnv
```

..  code-block:: typoscript
    :caption: EXT:site_package/Configuration/TypoScript/setup.typoscript

    lib.foo.data = GP : stuff

Get input value from query string `&stuff[bar]=...`

..  code-block:: typoscript
    :caption: EXT:site_package/Configuration/TypoScript/setup.typoscript

    lib.foo.data = GP : stuff|bar

..  _data-type-gettext-level:

level
-----

..  confval:: level
    :name: data-level

    Get the root line level of the current page.

..  _data-type-gettext-level-example:

Example: Get the root line level of the current page
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

..  code-block:: typoscript
    :caption: EXT:site_package/Configuration/TypoScript/setup.typoscript

    lib.foo.data = level

..  _data-type-gettext-levelfield:

levelfield
----------

..  confval:: levelfield
    :name: data-levelfield
    :Syntax: levelfield : [pointer, integer], [field name], ["slide"]

    Like :ref:`data-type-gettext-leveltitle` but the field do be fetched from the
    record is configurable.

..  _data-type-gettext-levelfield-example:

Example: Get a field from a page up in the root-line
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Get the value of the user defined field :sql:`tx_myextension_myfield` in the root line.

Reviewer: ```suggestion
Get the value of the user-defined field :sql:`tx_myextension_myfield` in the root line.
```

..  code-block:: typoscript
    :caption: EXT:my_extension/Configuration/TypoScript/setup.typoscript

    lib.foo.data = levelfield : -1, tx_myextension_myfield, slide

Requires additional configuration in

> :php:`$GLOBALS['TYPO3_CONF_VARS']['FE']['addRootLineFields']` to include

Reviewer: ```suggestion
:ref:`$GLOBALS['TYPO3_CONF_VARS']['FE']['addRootLineFields'] <t3coreapi:typo3ConfVars_fe_addRootLineFields>` to include
```

> field. In order that you can use this function, your field-name

Reviewer: ```suggestion
field. In order that you can use this function, your field name
```

..  _data-type-gettext-leveltitle:

leveltitle
----------

..  confval:: leveltitle
    :name: data-leveltitle
    :Syntax: leveltitle : [pointer, integer], ["slide"]

    Get the title of a page in the root-line.

    *   Use an absolute level with 0 or a positive integer.
    *   With a negative integer got x levels up
    *   The slide parameter slides until there is a non-empty value found.

Example: Get the title of a page up in the rootline
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Get the title of the page on the first level of the root line:

..  code-block:: typoscript
    :caption: EXT:site_package/Configuration/TypoScript/setup.typoscript

    lib.foo.data = leveltitle : 1

Get the title of the page on the level right below the current page AND if
that is not present, walk to the bottom of the root line until there's a
title:

..  code-block:: typoscript
    :caption: EXT:site_package/Configuration/TypoScript/setup.typoscript

    lib.foo.data = leveltitle : -2, slide

..  _data-type-gettext-leveluid:

leveluid
--------

..  confval:: leveluid
    :name: data-leveluid
    :Syntax: leveluid : [pointer, integer]

>     Get the UID of a page in the root-line.

Reviewer: ```suggestion
    Get the UID of a page in the root line.
```

    *   Use an absolute level with 0 or a positive integer.
    *   With a negative integer got x levels up

..  _data-type-gettext-leveluid-example:

> Example: Get the id of the root-page of the page-tree

Reviewer: ```suggestion
Example: Get the ID of the root page of the page tree
```

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Get the id of the root-page of the website (level zero):

Reviewer: ```suggestion
Get the ID of the root page of the website (level zero):
```

..  code-block:: typoscript
    :caption: EXT:site_package/Configuration/TypoScript/setup.typoscript

    lib.foo.data = leveluid : 0

..  _data-type-gettext-lll:

LLL
---

..  confval:: LLL
    :name: data-lll
    :Syntax: LLL: [fileref]:[labelkey]

>     Reference to a locallang (xlf, xml or php) label.  Reference consists of

Reviewer: ```suggestion
    Reference to a locallang label (:ref:`XLIFF <t3coreapi:xliff>`). Reference consists of
```

Only XLIFF left.

----

..  confval:: path
    :name: data-path

    Path to a file, possibly placed in an extension, returns empty if the file
    does not exist.

..  _data-type-gettext-path-example:

Example: Resolve the path to a file
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Get path to file rsaauth.js (inside extension rsaauth) relative to siteroot:

..  code-block:: typoscript
    :caption: EXT:site_package/Configuration/TypoScript/setup.typoscript

    lib.foo.data = path : EXT:rsaauth/resources/rsaauth.js

It can also be helpful in combination with the stdWrap function
:ref:`stdwrap-insertData`:

..  code-block:: typoscript
    :caption: EXT:site_package/Configuration/TypoScript/setup.typoscript

    page.headerData.10 = TEXT
    page.headerData.10 {
        data = path : EXT:site/Resources/Public/Images/logo.png
        wrap = <a href="|">
    }

..  _data-type-gettext-register:

register
--------

..  confval:: register
    :name: data-register
    :Syntax: register: [field name from the :php:`$GLOBALS['TSFE']->register`]

    See :ref:`LOAD_REGISTER <cobj-load-register>`.

..  _data-type-gettext-register-example:

Example: Get the content of a register
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Get content from :php:`$GLOBALS['TSFE']->register['color']`:

Reviewer: ```suggestion
Get content from the :ref:`register <using-setting-register>`:
```
$GLOBALS['TSFE']->register is now marked as internal, so we should not mention it:
https://github.com/TYPO3/typo3/blob/main/typo3/sysext/frontend/Classes/Controller/TypoScriptFrontendController.php#L177-L180

    Migrate these properties to use the config property. You can access the
    TypoScript properties directly, for example, via
    :typoscript:`lib.something.data = TSFE:config|config|fileTarget`

..  _data-type-gettext-tsfe-example:

Example: Get the username field of the current frontend user
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Get the "username" field of the current FE user:

Reviewer: Duplicated, can be removed.
